File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.cpp
class Solution {
public:
    int binarySearch(vector<int>& nums, int target) {
        int lo = 0;
        int hi = nums.size() - 1;
        while(lo < hi) {
            int mid = lo + (hi-lo) / 2;
            if(nums[mid] == target) return mid;
            else if(nums[mid] < target) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }
    vector<int> sortedSquares(vector<int>& nums) {
        vector<int> res(nums.size());
        int right = binarySearch(nums, 0);
        int left = right - 1;
        int i = 0;
        while(left >= 0 || right < nums.size()) {
            if(left < 0) {
                res[i] = nums[right] * nums[right];
                i++;
                right++;
            }
            else if(right >= nums.size()) {
                res[i] = nums[left] * nums[left];
                i++;
                left--;
            }
            else if(nums[right] < -nums[left]) {
                res[i] = nums[right] * nums[right];
                i++;
                right++;
            }
            else {
                res[i] = nums[left] * nums[left];
                i++;
                left--;
            }
        }
        return res;
    }
};

File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.py
class Solution:
    def binarySearch(self, nums: List[int], target: int) -> int:
        lo = 0
        hi = len(nums) - 1
        while lo <= hi:
            mid = lo + (hi-lo) // 2
            if nums[mid] == target: return mid
            elif nums[mid] < target: lo = mid + 1
            else: hi = mid - 1
        return lo
    def sortedSquares(self, nums: List[int]) -> List[int]:
        res = [0 for i in range(len(nums))]
        right = self.binarySearch(nums, 0)
        left = right - 1
        i = 0
        while left >= 0 or right < len(nums):
            if left < 0:
                res[i] = nums[right] * nums[right]
                i += 1
                right += 1
            elif right >= len(nums):
                res[i] = nums[left] * nums[left]
                i += 1
                left -= 1
            elif nums[right] < -nums[left]:
                res[i] = nums[right] * nums[right]
                i += 1
                right += 1
            else:
                res[i] = nums[left] * nums[left]
                i += 1
                left -= 1
        return res